core/src/main/java/org/apache/calcite/sql/SqlNumberTypeNameSpec.java

```diff
@@ -0,0 +1,120 @@
+
+/**
+ * A SQL type name specification for the NUMBER data type.
+ */
+public class SqlNumberTypeNameSpec extends SqlTypeNameSpec {
+
+  public final boolean isPrecisionStar;
+  public final SqlLiteral precision;
+  public final SqlLiteral scale;
+
+  /**
+   * Create a {@code SqlNumberTypeNameSpec} instance.
+   *
+   * @param isPrecisionStar Whether the precision element is the symbol "*"
+   * @param precision The precision, an unsigned integer in the range [1-38].
+   *                  May be null
+   * @param scale The scale, an unsigned integer in the range [0-38] if
+   *              isPrecisionStar = true. If isPrecisionStar = false, the valid
+   *              range is [0, precision]. May be null
+   * @param pos The parser position
+   */
+  public SqlNumberTypeNameSpec(boolean isPrecisionStar, SqlLiteral precision,
+      SqlLiteral scale, SqlParserPos pos) {
+    super(new SqlIdentifier("NUMBER", pos), pos);
+    this.isPrecisionStar = isPrecisionStar;
+    this.precision = precision;
+    this.scale = scale;
+    if (!isValidPrecision(isPrecisionStar, precision)) {
+      throw SqlUtil.newContextException(precision.getParserPosition(),
+          RESOURCE.numberLiteralOutOfRange(String.valueOf(precision)));
+    }
+    if (!isValidScale(isPrecisionStar, precision, scale)) {
+      throw SqlUtil.newContextException(scale.getParserPosition(),
+          RESOURCE.numberLiteralOutOfRange(String.valueOf(scale)));
+    }
+  }
+
+  private boolean isValidPrecision(boolean isPrecisionStar,
+      SqlLiteral precision) {
+    if (isPrecisionStar || precision == null) {
+      return true;
+    }
+    int numericPrecision = precision.getValueAs(Integer.class);
+    return numericPrecision >= 1 && numericPrecision <= 38;
+  }
+
+  private boolean isValidScale(boolean isPrecisionStar, SqlLiteral precision,
+      SqlLiteral scale) {
+    if (scale == null) {
+      return true;
+    }
+    int numericScale = scale.getValueAs(Integer.class);
+    if (isPrecisionStar) {
+      return numericScale >= 0 && numericScale <= 38;
+    } else {
+      int numericPrecision = precision.getValueAs(Integer.class);
+      return numericScale >= 0 && numericScale <= numericPrecision;
+    }
+  }
+
+  @Override public RelDataType deriveType(SqlValidator validator) {
+    return validator.getValidatedNodeType(getTypeName());
+  }
+
+  @Override public void unparse(SqlWriter writer, int leftPrec, int rightPrec) {
+    writer.keyword("NUMBER");
+    if (!isPrecisionStar && precision == null) {
+      return;
+    }
+    writer.setNeedWhitespace(false);
+    final SqlWriter.Frame frame = writer.startList("(", ")");
+    if (isPrecisionStar) {
+      writer.print("*");
+    } else {
+      precision.unparse(writer, 0, 0);
+    }
+    if (scale != null) {
+      writer.sep(",", true);
+      scale.unparse(writer, 0, 0);
+    }
+    writer.endList(frame);
+  }
+
+  @Override public boolean equalsDeep(SqlTypeNameSpec spec, Litmus litmus) {
```

Comment:
What happens if we just replace all of this logic with return false? Do any tests fail?

If not, I'd suggest just doing that. It's usually better to have no logic at all than logic that's untested.

Reply:
Yeah it still passes all the tests, I'll change it to just false then